File: go/kswitch/deny_reason.go
// Package kswitch — DenyReason (EP-050 W4 parity with server).
//
// Wire parity: matches app/enforcement/reason_class.py::DenyReason (server),
// sdks/python/kswitch/deny_reason.py (Python), and
// sdks/typescript/src/denyReason.ts (TypeScript).
//
// Forward compatibility: ParseDenyReason returns DenyReasonUnknown for any
// value this SDK does not yet recognise — never panics.

package kswitch

import "strings"

// DenyReason is the semantic category of an enforcement deny decision.
type DenyReason string

const (
	DenyReasonPolicy      DenyReason = "POLICY"
	DenyReasonGovernance  DenyReason = "GOVERNANCE"
	DenyReasonUnavailable DenyReason = "UNAVAILABLE"
	DenyReasonValidation  DenyReason = "VALIDATION"
	DenyReasonUnknown     DenyReason = "UNKNOWN"
)

// denyReasonKnown is the set of values this SDK recognises.  Keep in sync
// with the constants above.
var denyReasonKnown = map[DenyReason]struct{}{
	DenyReasonPolicy:      {},
	DenyReasonGovernance:  {},
	DenyReasonUnavailable: {},
	DenyReasonValidation:  {},
	DenyReasonUnknown:     {},
}

// ParseDenyReason parses a string into a DenyReason.  Unknown values return
// DenyReasonUnknown rather than an error — enforcement callers should never
// fail-closed on a *category* mismatch (the wire-level `decision` field is
// the source of truth; this is metadata).
func ParseDenyReason(raw string) DenyReason {
	if raw == "" {
		return DenyReasonUnknown
	}
	v := DenyReason(strings.ToUpper(strings.TrimSpace(raw)))
	if _, ok := denyReasonKnown[v]; ok {
		return v
	}
	return DenyReasonUnknown
}

// String implements fmt.Stringer.
func (r DenyReason) String() string { return string(r) }

File: python/kswitch/deny_reason.py
"""KSwitch Python SDK — DenyReason enum (EP-050 W4 parity with server).

Usage::

    from kswitch.deny_reason import DenyReason, parse_deny_reason

    deny = parse_deny_reason(response.get("deny_reason"))
    if deny is DenyReason.UNAVAILABLE:
        # control-plane is down — back off and retry
        ...

Forward compatibility: unknown server values fall through to
``DenyReason.UNKNOWN`` rather than raising.  This lets the server roll out
new categories before every SDK pins a matching release.

Wire parity
-----------
Matches ``app/enforcement/reason_class.py::DenyReason`` (server).  The five
values are intentionally a small, stable set; adding a new value requires a
coordinated server-SDK rollout documented in the EP-050 closure EP.
"""
from __future__ import annotations

import enum
from typing import Any


class DenyReason(str, enum.Enum):
    """Semantic category of an enforcement deny decision.

    String-valued — serialises directly to JSON.
    """

    POLICY = "POLICY"
    GOVERNANCE = "GOVERNANCE"
    UNAVAILABLE = "UNAVAILABLE"
    VALIDATION = "VALIDATION"
    UNKNOWN = "UNKNOWN"

    @classmethod
    def parse(cls, raw: Any) -> "DenyReason":
        """Parse any input into a DenyReason; unknown values → UNKNOWN."""
        if isinstance(raw, cls):
            return raw
        if not isinstance(raw, str):
            return cls.UNKNOWN
        try:
            return cls(raw.upper())
        except ValueError:
            return cls.UNKNOWN


def parse_deny_reason(raw: Any) -> DenyReason:
    """Module-level convenience alias for :meth:`DenyReason.parse`."""
    return DenyReason.parse(raw)


__all__ = ["DenyReason", "parse_deny_reason"]
